### DISCUSSION TITLE
Help connecting to HID Signo 40 Reader

### DISCUSSION THREAD

**Discussion by irisi-jr** [goToMain/libosdp#175] [General]
Hi there,

This is likely a problem with my configuration and understanding (i am new to python and card readers), I am hoping for a little guidance, and not sure where to ask.. sorry if this is the wrong place!

I am attempting to communicate with a HID Signo 40 reader over rs485 on my machine, but i am having no luck with the example python script (cp_app.py)

when I run the script - ./cp_app.py /dev/ttyS0 
```text
pyosdp: [2024-04-18T01:30:49Z] [INFO ] vendor/src/osdp_cp.c:1372: Setup complete; LibOSDP-3.0.5 master (bd15c2b) NumPDs: 1 Channels: 1
OSDP: CP: PD-101: [2024-04-18T01:30:50Z] [ERROR] vendor/src/osdp_cp.c:890: Response timeout for CMD(61)
```
I've configured the reader in install mode over osdp v2. 

I suspect i am not understanding how to configure the test script correctly, i am hoping for some pointers.

EG
```py
pd_info = [
        PDInfo(101, channel, scbk=KeyStore.gen_key()),
]
```
does '101' need to be changed to something else (ie my reader details)? if so how do i find out what it needs to be?

Any other general guidance would be greatly appreciated.

Thanks,

**Comment by sidcha** [comment]
Hey!

> [ERROR] vendor/src/osdp_cp.c:890: Response timeout for CMD(61)

This line means that your CP (the python program) sent a command to the PD and it didn't respond withing the given time span (200ms). A number of things can cause this:
- baudrate is incorrect
- the device /dev/ttyS0 isn't the one the PD is connected to.
- The the PD is wired incorrectly (RX and TX wires swapped?)
- Reader address is incorrect

> does '101' need to be changed to something else (ie my reader details)?

Yes, that has to be your reader's address.

**Comment by irisi-jr** [comment]
Thanks - I appreciate your response. I managed to get it talking and have a python script running as a cp sending commands/receiving events successfully.

I am having a bit of trouble understanding how to set up the secure channel properly.

At the moment I am running in install mode with scbk set to none and PD reader in 'install' mode - PDInfo(0, channel, None)

This works, in set up i get msg "Online; Without SC"

What is the correct process to install a key and then use that key for future start ups in SC mode? Do i need to install the key, for example, from a separate once off script, and then initiate the connection after that with the same key? 

Additionally is there an easy way to check from python whether the cp-pd connection started ok?

Thanks!

**Reply by sidcha** [reply]
You should not set SCBK to None in CP mode if you want to use SC; doing so will cause SC to be disabled. You have to set the PD to install mode and do:

```py
store = KeyStore("/some/path/to/key/store")
pd_info = [
        PDInfo(101, channel, scbk=store.new_key("pd-101")),
]
cp = ControlPanel(pd_info)
```
To use a previously stored key, instead of `store.new_key()` do `store.load_key())`.

**Comment by dmercer-google** [comment]
Looks like you may be generating a random SCP key to connect. This won't work as both the reader and CP must know the same key to communicate. If your reader has no key set you can just send null. If it has a key you need to know what it is. You can try the SCP default key '0x404142434445464748494A4B4C4D4E4F'

To change the key send the OSDP set key command under a channel that is encrypted using the existing key or not encrypted if there is none.

**Reply by sidcha** [reply]
The default key from the spec is an implementation detail and is abstracted away by LibOSDP. Users don't need to manually set their key to the default key ever (actually, it's wrong to do so; we don't want another [Black Hat talk](https://youtu.be/0yrHtJY3mww) trolling us).

If you are CP interacting with a PD there are 2 cases:

- you know the SCBK set in the PD (from a past successful keyset); you pass that during CP setup and a secure channel will be established.

- you don't know the SCBK (new device?); you set the PD to "install-mode" by whichever mechanism the PD vendor documents and then generate a new key and call CP setup with that key. Now LibOSDP will detect that the PD is in "install-mode" and will setup a temporary SC session with the default key and use it to set the new key provided during setup. After this, CP app must persist this key in some non-volatile storage and use it henceforth when calling cp_setup for that PD.

Please note, "install-mode" is an insecure operation mode and must be performed under supervision out of proportion scenarios (ie, the PD must be connected directly to a CP by a short wire).

**Reply by dmercer-google** [reply]
Thanks for the clarification.  I mentioned the default key because I have received readers in the past that have the 40...4F key set as default.  No idea why, its just how they came.

**Reply by sidcha** [reply]
Just to clarify, all PD must be shipped with the default key by design. It's just that LibOSDP deals with this implementation detail so users don't need to care about it. They just need to worry about PD being in install-mode.

**Reply by irisi-jr** [reply]
thanks to both of you for the assistance, very helpful!
